File: assembler/construct_snarl_tree_map.py
"""
Constructs a hierarchical snarl tree map from a variation graph.

This script traverses a snarl decomposition of a variation graph, calculating
the number of leaf snarls in the subtree of each snarl and chain. It supports
parallel processing for large graphs by dividing the traversal work at a
predefined depth in the tree.

The output is a JSON file representing the snarl tree, where each key is a
snarl or chain ID and the value contains its children and the total count
of leaf snarls in its subtree.
"""
import os
import multiprocessing
from bdsg.bdsg import PackedGraph
from bdsg.bdsg import SnarlDistanceIndex
import time
import json
from sys import stderr
import argparse

PARALLELIZATION_DEPTH = 2

g_snarl_decomposition_obj = None

def init_worker(snarl_decomposition_obj):
    """
    Initializes each worker process for parallel execution.

    This function is called by multiprocessing.Pool to set up the global
    SnarlDecomposition object in each worker, allowing for memory sharing
    of the loaded graph and index via copy-on-write.
    """
    global g_snarl_decomposition_obj
    g_snarl_decomposition_obj = snarl_decomposition_obj

def process_subtree_worker(args):
    """
    Performs snarl tree traversal on a specific subtree.

    This is the main worker function called by the multiprocessing pool. It
    uses the globally shared SnarlDecomposition object to process a single
    subtree and returns the resulting map.
    """
    worker_id, subtree_root_handle = args
    start_time = time.time()
        
    result = g_snarl_decomposition_obj.traverse_decomposition(subtree_root_handle)
        
    end_time = time.time()
    print(f"[Worker {worker_id}] Finished in {end_time - start_time:.2f} seconds.", file=stderr)
        
    return result

class SnarlDecomposition:
    """
    Manages the decomposition of a variation graph into a snarl tree.

    This class encapsulates the graph and its distance index, providing methods
    to traverse the snarl decomposition and build a hierarchical map.
    """

    def __init__(self):
        """Initializes the SnarlDecomposition object."""
        self.graph = PackedGraph()
        self.index = SnarlDistanceIndex()
        self.snarl_tree_map = {}   # snarl_tree_map[snarl_id] = {[children_ids], count_of_leaf_snarls_in_subtree}

    def build(self, graph_path, index_path):
        """Loads the graph and snarl distance index from disk."""
        t0 = time.time()
        self.graph.deserialize(graph_path)
        print(f"Graph deserialized in {time.time()-t0:.2f}", file=stderr, flush=True)
        self.index.deserialize(index_path)
        print(f"Index deserialized in {time.time()-t0:.2f}", file=stderr, flush=True)


    def get_id(self, net):
        """Generates a unique, human-readable ID for a net object."""
        if self.index.is_chain(net) or self.index.is_snarl(net):
            left_bound_net = self.index.get_bound(net, False, False)
            right_bound_net = self.index.get_bound(net, True, False)
            left_bound_node_id = self.graph.get_id(self.index.get_handle(left_bound_net, self.graph))
            right_bound_node_id = self.graph.get_id(self.index.get_handle(right_bound_net, self.graph))
            if left_bound_node_id < right_bound_node_id:
                left_bound_node_id, right_bound_node_id = right_bound_node_id, left_bound_node_id
            bound_str = str(left_bound_node_id) + "-" + str(right_bound_node_id)
            if self.index.is_chain(net):
                return "c" + bound_str
            else:
                return "s" + bound_str
        elif self.index.is_root(net):
            return "root"
        elif self.index.is_node(net):
            return str(self.graph.get_id(self.index.get_handle(net, self.graph)))
        else:
            raise ValueError(f"Net {net} is not a chain, snarl, node or root")

    def check_snarl_in_children_iteratee(self, child_net_handle) -> bool:
        """Callback function to check if a net has snarls as children."""
        if self.index.is_snarl(child_net_handle):
            self.contains_child_snarls = True
            return False
        return True

    def is_leaf_snarl(self, net):
        """Determines if a snarl is a leaf in the snarl tree."""
        self.contains_child_snarls = False

        snarl_children: list = []
        self.index.for_each_child(
            net, lambda y: snarl_children.append(y) or True
        )

        for snarl_child in snarl_children:
            self.index.for_each_child(snarl_child, self.check_snarl_in_children_iteratee)

        if not self.contains_child_snarls:
            return True

    
    def traverse_decomposition_helper(self, handle, snarl_tree_map):
        """
        Recursively traverses a subtree to count leaf snarls.

        This is the core recursive function for building the snarl tree map.
        It performs a depth-first traversal, counting leaf snarls and
        propagating the counts up the tree.
        """
        map_key = self.get_id(handle)
        
        # The stopping condition is when the current net is a leaf snarl
        if self.index.is_snarl(handle) and self.is_leaf_snarl(handle):
            snarl_tree_map[map_key] = ([], 1)
            return 1, map_key
        elif self.index.is_node(handle):
            return 0, map_key

        cnt_leaf_snarls_in_subtree = 0
        children_handles = []
        children_map_keys = []
        self.index.for_each_child(handle, lambda y: (children_handles.append(y) if not self.index.is_node(y) else True) or True)
        for child_handle in children_handles:
            cnt_leaf_snarls_in_child_subtree, child_map_key = self.traverse_decomposition_helper(child_handle, snarl_tree_map)
            children_map_keys.append(child_map_key)
            # Once the recursive call for a child returns, it will have returned the total number of leaf snarls in that child's entire subtree. This count is added to the current snarl's running total.
            cnt_leaf_snarls_in_subtree += cnt_leaf_snarls_in_child_subtree
        
        # After the for loop ends, the function has visited all the current snarl's children and has received the total leaf snarl counts for each of them.
        snarl_tree_map[map_key] = (children_map_keys, cnt_leaf_snarls_in_subtree)
        
        return cnt_leaf_snarls_in_subtree, map_key


    def traverse_decomposition(self, subtree_root_handle):
        """
        Traverses a subtree to build a snarl tree map.

        This function is re-entrant and can be called from multiple processes.
        It initializes a new map for each traversal and returns it.
        """
        snarl_tree_map = {}
        self.traverse_decomposition_helper(subtree_root_handle, snarl_tree_map)
        # Handle the case where the root of the subtree is itself a leaf snarl
        if not snarl_tree_map and self.index.is_snarl(subtree_root_handle) and self.is_leaf_snarl(subtree_root_handle):
             map_key = self.get_id(subtree_root_handle)
             snarl_tree_map[map_key] = ([], 1)
        return snarl_tree_map


    def collect_handles_at_depth(self, depth):
        """
        Collects all snarl and chain handles at a specific depth.

        This function now works by iterating through the nodes of parent
        handles to find the unique child snarls and chains.
        """
        if depth == 0:
            return [self.index.get_root()]

        parent_handles = self.collect_handles_at_depth(depth - 1)

        child_handles_at_parallelization_depth = []

        for parent_handle in parent_handles:
            self.index.for_each_child(parent_handle, lambda y: (child_handles_at_parallelization_depth.append(y) if not self.index.is_node(y) else True) or True)

        return child_handles_at_parallelization_depth


    def build_top_tree(self, handle, current_depth):
        """
        Recursively reconstructs the top levels of the snarl tree map.

        After parallel processing of the lower levels, this function is called
        on the root to rebuild the top of the tree, integrating the results
        from the worker processes.
        """
        if current_depth >= PARALLELIZATION_DEPTH:
            return self.snarl_tree_map.get(self.get_id(handle), ([], 0))[1]

        map_key = self.get_id(handle)
        
        if map_key in self.snarl_tree_map and self.snarl_tree_map[map_key][1] > 0:
            return self.snarl_tree_map[map_key][1]

        children_handles = []
        self.index.for_each_child(handle, lambda y: (children_handles.append(y) if not self.index.is_node(y) else True) or True)

        children_map_keys = []
        total_leaf_snarls = 0

        for child_handle in children_handles:
            children_map_keys.append(self.get_id(child_handle))
            count = self.build_top_tree(child_handle, current_depth + 1)
            total_leaf_snarls += count

        if not children_handles and self.index.is_snarl(handle) and self.is_leaf_snarl(handle):
            total_leaf_snarls = 1

        self.snarl_tree_map[map_key] = (children_map_keys, total_leaf_snarls)
        return total_leaf_snarls


def main():
    """
    Main function to orchestrate the snarl tree map construction.

    Parses command-line arguments, loads the graph and index, manages the
    parallel processing of subtrees, reconstructs the final tree, and saves
    the result to a JSON file.
    """
    parser = argparse.ArgumentParser(description="Construct a snarl tree map from a variation graph.")
    parser.add_argument("-g", "--graph", required=True, help="Path to the variation graph (.pg file).")
    parser.add_argument("-i", "--index", required=True, help="Path to the snarl distance index (.dist file).")
    parser.add_argument("-o", "--output-json", default="snarl_tree_map.json", help="Path to the output JSON file.")
    args = parser.parse_args()
    
    print("Starting parallel snarl tree map construction.", file=stderr)

    main_obj = SnarlDecomposition()
    main_obj.build(args.graph, args.index)

    handles_to_process = main_obj.collect_handles_at_depth(PARALLELIZATION_DEPTH)
    print(f"Found {len(handles_to_process)} handles to process.", file=stderr)
    
    num_processes = min(multiprocessing.cpu_count(), len(handles_to_process)) if handles_to_process else 0
    
    print(f"Running with {num_processes} parallel processes.", file=stderr)

    if num_processes > 0:
        # We use the 'fork' start method (default on Linux) to allow for memory
        # sharing of the main_obj between processes (copy-on-write).
        with multiprocessing.Pool(processes=num_processes, initializer=init_worker, initargs=(main_obj,)) as pool:
            worker_args = [(i, handle) for i, handle in enumerate(handles_to_process)]
            results = pool.map(process_subtree_worker, worker_args)
            
            merged_map = {}
            for result_map in results:
                merged_map.update(result_map)
            
            # The main object's map now contains the results from all workers
            main_obj.snarl_tree_map = merged_map
    else: # If no handles to process, the tree is smaller than PARALLELIZATION_DEPTH
        # If the tree is very small, just traverse it sequentially.
        main_obj.snarl_tree_map = main_obj.traverse_decomposition(main_obj.index.get_root())

    # Reconstruct the top levels of the tree using the processed subtrees
    main_obj.build_top_tree(main_obj.index.get_root(), 0)
    
    # save to json
    with open(args.output_json, "w") as f:
        json.dump(main_obj.snarl_tree_map, f, indent=4)

if __name__ == "__main__":
    main()

File: snarl_tree_builder_cpp/construct_snarl_tree_map.cpp
/**
 * Constructs a hierarchical snarl tree map from a variation graph.
 * 
 * This program traverses a snarl decomposition of a variation graph, calculating
 * the number of leaf snarls in the subtree of each snarl and chain. It supports
 * parallel processing for large graphs using C++ multithreading.
 * 
 * The output is a JSON file representing the snarl tree, where each key is a
 * snarl or chain ID and the value contains its children and the total count
 * of leaf snarls in its subtree.
 */

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <thread>
#include <mutex>
#include <chrono>
#include <algorithm>
#include <memory>
#include <sstream>
#include <iomanip>

#include "bdsg/packed_graph.hpp"
#include "bdsg/snarl_distance_index.hpp"
#include "handlegraph/types.hpp"
#include "handlegraph/util.hpp"

using namespace std;
using namespace bdsg;
using namespace handlegraph;

const int PARALLELIZATION_DEPTH = 2;

// Structure to hold the tree map entry: (children_ids, leaf_count)
struct TreeMapEntry {
    vector<string> children_ids;
    int leaf_count;
};

class SnarlTreeBuilder {
private:
    PackedGraph graph;
    SnarlDistanceIndex index;
    map<string, TreeMapEntry> snarl_tree_map;
    mutex map_mutex;  // For thread-safe map operations
    
    /**
     * Generates a unique, human-readable ID for a net object.
     */
    string get_id(const net_handle_t& net) {
        if (index.is_chain(net) || index.is_snarl(net)) {
            net_handle_t left_bound_net = index.get_bound(net, false, false);
            net_handle_t right_bound_net = index.get_bound(net, true, false);
            nid_t left_bound_node_id = graph.get_id(index.get_handle(left_bound_net, &graph));
            nid_t right_bound_node_id = graph.get_id(index.get_handle(right_bound_net, &graph));
            
            if (left_bound_node_id < right_bound_node_id) {
                swap(left_bound_node_id, right_bound_node_id);
            }
            
            string bound_str = to_string(left_bound_node_id) + "-" + to_string(right_bound_node_id);
            
            if (index.is_chain(net)) {
                return "c" + bound_str;
            } else {
                return "s" + bound_str;
            }
        } else if (index.is_root(net)) {
            return "root";
        } else if (index.is_node(net)) {
            return to_string(graph.get_id(index.get_handle(net, &graph)));
        } else {
            throw runtime_error("Net is not a chain, snarl, node or root");
        }
    }
    
    /**
     * Determines if a snarl is a leaf in the snarl tree.
     */
    bool is_leaf_snarl(const net_handle_t& net) {
        bool contains_child_snarls = false;
        
        vector<net_handle_t> snarl_children;
        index.for_each_child(net, [&](const net_handle_t& child) {
            snarl_children.push_back(child);
        });
        
        for (const auto& snarl_child : snarl_children) {
            index.for_each_child(snarl_child, [&](const net_handle_t& grandchild) {
                if (index.is_snarl(grandchild)) {
                    contains_child_snarls = true;
                    return false;  // Stop iteration
                }
                return true;
            });
            if (contains_child_snarls) {
                break;
            }
        }
        
        return !contains_child_snarls;
    }
    
    /**
     * Recursively traverses a subtree to count leaf snarls.
     */
    pair<int, string> traverse_decomposition_helper(const net_handle_t& handle, 
                                                     map<string, TreeMapEntry>& local_map) {
        string map_key = get_id(handle);
        
        // Stopping condition: leaf snarl
        if (index.is_snarl(handle) && is_leaf_snarl(handle)) {
            local_map[map_key] = {{}, 1};
            return {1, map_key};
        } else if (index.is_node(handle)) {
            return {0, map_key};
        }
        
        int cnt_leaf_snarls_in_subtree = 0;
        vector<net_handle_t> children_handles;
        vector<string> children_map_keys;
        
        // Collect non-node children
        index.for_each_child(handle, [&](const net_handle_t& child) {
            if (!index.is_node(child)) {
                children_handles.push_back(child);
            }
        });
        
        for (const auto& child_handle : children_handles) {
            auto [cnt_leaf_snarls_in_child, child_map_key] = 
                traverse_decomposition_helper(child_handle, local_map);
            children_map_keys.push_back(child_map_key);
            cnt_leaf_snarls_in_subtree += cnt_leaf_snarls_in_child;
        }
        
        local_map[map_key] = {children_map_keys, cnt_leaf_snarls_in_subtree};
        
        return {cnt_leaf_snarls_in_subtree, map_key};
    }
    
    /**
     * Traverses a subtree to build a snarl tree map.
     * Thread-safe for parallel execution.
     */
    map<string, TreeMapEntry> traverse_decomposition(const net_handle_t& subtree_root_handle) {
        map<string, TreeMapEntry> local_map;
        traverse_decomposition_helper(subtree_root_handle, local_map);
        
        // Handle the case where the root is itself a leaf snarl
        if (local_map.empty() && index.is_snarl(subtree_root_handle) && 
            is_leaf_snarl(subtree_root_handle)) {
            string map_key = get_id(subtree_root_handle);
            local_map[map_key] = {{}, 1};
        }
        
        return local_map;
    }
    
    /**
     * Collects all snarl and chain handles at a specific depth.
     */
    vector<net_handle_t> collect_handles_at_depth(int depth) {
        if (depth == 0) {
            return {index.get_root()};
        }
        
        vector<net_handle_t> parent_handles = collect_handles_at_depth(depth - 1);
        vector<net_handle_t> child_handles;
        
        for (const auto& parent_handle : parent_handles) {
            index.for_each_child(parent_handle, [&](const net_handle_t& child) {
                if (!index.is_node(child)) {
                    child_handles.push_back(child);
                }
            });
        }
        
        return child_handles;
    }
    
    /**
     * Worker function for parallel processing.
     */
    void process_subtree_worker(int worker_id, const net_handle_t& handle) {
        auto local_map = traverse_decomposition(handle);
        
        // Merge into shared map (thread-safe)
        lock_guard<mutex> lock(map_mutex);
        snarl_tree_map.insert(local_map.begin(), local_map.end());
    }
    
    /**
     * Recursively reconstructs the top levels of the snarl tree map.
     */
    int build_top_tree(const net_handle_t& handle, int current_depth) {
        if (current_depth >= PARALLELIZATION_DEPTH) {
            string key = get_id(handle);
            auto it = snarl_tree_map.find(key);
            if (it != snarl_tree_map.end()) {
                return it->second.leaf_count;
            }
            return 0;
        }
        
        string map_key = get_id(handle);
        
        // Check if already computed
        auto it = snarl_tree_map.find(map_key);
        if (it != snarl_tree_map.end() && it->second.leaf_count > 0) {
            return it->second.leaf_count;
        }
        
        vector<net_handle_t> children_handles;
        index.for_each_child(handle, [&](const net_handle_t& child) {
            if (!index.is_node(child)) {
                children_handles.push_back(child);
            }
        });
        
        vector<string> children_map_keys;
        int total_leaf_snarls = 0;
        
        for (const auto& child_handle : children_handles) {
            children_map_keys.push_back(get_id(child_handle));
            int count = build_top_tree(child_handle, current_depth + 1);
            total_leaf_snarls += count;
        }
        
        if (children_handles.empty() && index.is_snarl(handle) && is_leaf_snarl(handle)) {
            total_leaf_snarls = 1;
        }
        
        snarl_tree_map[map_key] = {children_map_keys, total_leaf_snarls};
        return total_leaf_snarls;
    }
    
    /**
     * Escapes a string for JSON output.
     */
    string json_escape(const string& s) {
        ostringstream o;
        for (auto c : s) {
            switch (c) {
                case '"': o << "\\\""; break;
                case '\\': o << "\\\\"; break;
                case '\b': o << "\\b"; break;
                case '\f': o << "\\f"; break;
                case '\n': o << "\\n"; break;
                case '\r': o << "\\r"; break;
                case '\t': o << "\\t"; break;
                default:
                    if ('\x00' <= c && c <= '\x1f') {
                        o << "\\u" << hex << setw(4) << setfill('0') << (int)c;
                    } else {
                        o << c;
                    }
            }
        }
        return o.str();
    }
    
    /**
     * Writes the snarl tree map to a JSON file.
     */
    void write_json(const string& output_path) {
        ofstream out(output_path);
        if (!out.is_open()) {
            throw runtime_error("Cannot open output file: " + output_path);
        }
        
        out << "{\n";
        bool first = true;
        for (const auto& [key, entry] : snarl_tree_map) {
            if (!first) {
                out << ",\n";
            }
            first = false;
            
            out << "    \"" << json_escape(key) << "\": [\n";
            out << "        [\n";
            
            // Write children IDs
            for (size_t i = 0; i < entry.children_ids.size(); ++i) {
                out << "            \"" << json_escape(entry.children_ids[i]) << "\"";
                if (i < entry.children_ids.size() - 1) {
                    out << ",";
                }
                out << "\n";
            }
            
            out << "        ],\n";
            out << "        " << entry.leaf_count << "\n";
            out << "    ]";
        }
        out << "\n}\n";
        
        out.close();
    }
    
public:
    /**
     * Loads the graph and snarl distance index from disk.
     */
    void load(const string& graph_path, const string& index_path) {
        auto t0 = chrono::high_resolution_clock::now();
        
        ifstream graph_file(graph_path);
        if (!graph_file.is_open()) {
            throw runtime_error("Cannot open graph file: " + graph_path);
        }
        graph.deserialize(graph_file);
        graph_file.close();
        
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed = t1 - t0;
        cerr << "Loading graph: " << fixed << setprecision(2) 
             << elapsed.count() << " seconds" << endl;
        
        ifstream index_file(index_path);
        if (!index_file.is_open()) {
            throw runtime_error("Cannot open index file: " + index_path);
        }
        index.deserialize(index_file);
        index_file.close();
        
        auto t2 = chrono::high_resolution_clock::now();
        elapsed = t2 - t1;
        cerr << "Loading index: " << fixed << setprecision(2) 
             << elapsed.count() << " seconds" << endl;
    }
    
    /**
     * Builds the snarl tree map using parallel threads.
     */
    void build_parallel(const string& output_path) {
        auto workflow_start = chrono::high_resolution_clock::now();
        
        vector<net_handle_t> handles_to_process = collect_handles_at_depth(PARALLELIZATION_DEPTH);
        
        unsigned int num_threads = min((unsigned int)thread::hardware_concurrency(), 
                                      (unsigned int)handles_to_process.size());
        if (num_threads == 0) num_threads = 1;
        
        cerr << "Running with " << num_threads << " parallel threads." << endl;
        
        if (!handles_to_process.empty()) {
            // Use a thread pool approach to avoid creating too many threads
            vector<thread> threads;
            size_t handles_per_thread = (handles_to_process.size() + num_threads - 1) / num_threads;
            
            for (size_t t = 0; t < num_threads; ++t) {
                threads.emplace_back([this, t, &handles_to_process, handles_per_thread, num_threads]() {
                    size_t start_idx = t * handles_per_thread;
                    size_t end_idx = min(start_idx + handles_per_thread, handles_to_process.size());
                    
                    for (size_t i = start_idx; i < end_idx; ++i) {
                        process_subtree_worker(i, handles_to_process[i]);
                    }
                });
            }
            
            // Wait for all threads to complete
            for (auto& thread : threads) {
                thread.join();
            }
        } else {
            // If no handles to process, traverse from root sequentially
            snarl_tree_map = traverse_decomposition(index.get_root());
        }
        
        // Reconstruct the top levels of the tree
        build_top_tree(index.get_root(), 0);
        
        // Write output
        write_json(output_path);
        
        auto workflow_end = chrono::high_resolution_clock::now();
        chrono::duration<double> total_elapsed = workflow_end - workflow_start;
        
        cerr << "Building snarl tree: " << fixed << setprecision(2) 
             << total_elapsed.count() << " seconds (produced " 
             << snarl_tree_map.size() << " entries)" << endl;
    }
};

int main(int argc, char* argv[]) {
    // Simple argument parsing
    string graph_path, index_path, output_path = "snarl_tree_map.json";
    
    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if ((arg == "-g" || arg == "--graph") && i + 1 < argc) {
            graph_path = argv[++i];
        } else if ((arg == "-i" || arg == "--index") && i + 1 < argc) {
            index_path = argv[++i];
        } else if ((arg == "-o" || arg == "--output-json") && i + 1 < argc) {
            output_path = argv[++i];
        } else if (arg == "-h" || arg == "--help") {
            cout << "Usage: " << argv[0] << " -g GRAPH -i INDEX [-o OUTPUT]\n";
            cout << "  -g, --graph       Path to the variation graph (.pg file)\n";
            cout << "  -i, --index       Path to the snarl distance index (.dist file)\n";
            cout << "  -o, --output-json Path to the output JSON file (default: snarl_tree_map.json)\n";
            return 0;
        }
    }
    
    if (graph_path.empty() || index_path.empty()) {
        cerr << "Error: Both -g/--graph and -i/--index are required.\n";
        cerr << "Use -h or --help for usage information.\n";
        return 1;
    }
    
    try {
        SnarlTreeBuilder decomp;
        decomp.load(graph_path, index_path);
        decomp.build_parallel(output_path);
    } catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }
    
    return 0;
}

